File: server/app/services/google_calendar.py
from datetime import datetime, timedelta, timezone

import httpx

CALENDAR_LIST_URL = "https://www.googleapis.com/calendar/v3/users/me/calendarList"
CALENDAR_EVENTS_URL = "https://www.googleapis.com/calendar/v3/calendars/{calendar_id}/events"


async def fetch_calendars(access_token: str) -> list[dict]:
    """Fetch the user's Google Calendar list."""
    headers = {"Authorization": f"Bearer {access_token}"}

    async with httpx.AsyncClient() as client:
        response = await client.get(CALENDAR_LIST_URL, headers=headers, timeout=10.0)
        response.raise_for_status()
        data = response.json()

    return [
        {
            "id": cal["id"],
            "name": cal.get("summary", ""),
            "color": cal.get("backgroundColor", ""),
        }
        for cal in data.get("items", [])
    ]


async def fetch_events(
    access_token: str,
    calendar_ids: list[str],
    days_ahead: int = 7,
) -> list[dict]:
    """Fetch events from one or more Google Calendars."""
    headers = {"Authorization": f"Bearer {access_token}"}
    now = datetime.now(timezone.utc)
    time_min = now.isoformat()
    time_max = (now + timedelta(days=days_ahead)).isoformat()

    all_events: list[dict] = []

    async with httpx.AsyncClient() as client:
        for calendar_id in calendar_ids:
            url = CALENDAR_EVENTS_URL.format(calendar_id=calendar_id)
            params = {
                "timeMin": time_min,
                "timeMax": time_max,
                "singleEvents": "true",
                "orderBy": "startTime",
            }
            response = await client.get(url, headers=headers, params=params, timeout=10.0)
            response.raise_for_status()
            data = response.json()

            for item in data.get("items", []):
                start = item.get("start", {})
                end = item.get("end", {})
                all_day = "date" in start and "dateTime" not in start

                all_events.append({
                    "title": item.get("summary", "(No title)"),
                    "start": start.get("date") if all_day else start.get("dateTime"),
                    "end": end.get("date") if all_day else end.get("dateTime"),
                    "calendar_id": calendar_id,
                    "color": item.get("colorId", ""),
                    "all_day": all_day,
                })

    return all_events

File: src/server/services/google-calendar.ts
const CALENDAR_LIST_URL = 'https://www.googleapis.com/calendar/v3/users/me/calendarList';
const CALENDAR_EVENTS_URL = 'https://www.googleapis.com/calendar/v3/calendars/{calendar_id}/events';

export interface Calendar {
  id: string;
  name: string;
  color: string;
}

export interface CalendarEvent {
  title: string;
  start: string;
  end: string;
  calendar_id: string;
  color: string;
  all_day: boolean;
}

export async function fetchCalendars(accessToken: string): Promise<Calendar[]> {
  const response = await fetch(CALENDAR_LIST_URL, {
    headers: { Authorization: `Bearer ${accessToken}` },
    signal: AbortSignal.timeout(10_000),
  });

  if (!response.ok) {
    throw new Error(`Google Calendar API error: ${response.status} ${response.statusText}`);
  }

  const data = await response.json();

  return (data.items ?? []).map((cal: Record<string, string>) => ({
    id: cal.id,
    name: cal.summary ?? '',
    color: cal.backgroundColor ?? '',
  }));
}

export async function fetchEvents(
  accessToken: string,
  calendarIds: string[],
  daysAhead: number = 7,
): Promise<CalendarEvent[]> {
  const now = new Date();
  const timeMin = now.toISOString();
  const timeMax = new Date(now.getTime() + daysAhead * 24 * 60 * 60 * 1000).toISOString();

  const allEvents: CalendarEvent[] = [];

  for (const calendarId of calendarIds) {
    const encodedId = encodeURIComponent(calendarId);
    const url = CALENDAR_EVENTS_URL.replace('{calendar_id}', encodedId);
    const params = new URLSearchParams({
      timeMin,
      timeMax,
      singleEvents: 'true',
      orderBy: 'startTime',
    });

    const response = await fetch(`${url}?${params.toString()}`, {
      headers: { Authorization: `Bearer ${accessToken}` },
      signal: AbortSignal.timeout(10_000),
    });

    if (!response.ok) {
      throw new Error(`Google Calendar API error: ${response.status} ${response.statusText}`);
    }

    const data = await response.json();

    for (const item of data.items ?? []) {
      const start = item.start ?? {};
      const end = item.end ?? {};
      const allDay = 'date' in start && !('dateTime' in start);

      allEvents.push({
        title: item.summary ?? '(No title)',
        start: allDay ? start.date : start.dateTime,
        end: allDay ? end.date : end.dateTime,
        calendar_id: calendarId,
        color: item.colorId ?? '',
        all_day: allDay,
      });
    }
  }

  return allEvents;
}
